node/pkg/common/queryResponse.go

```diff
@@ -34,72 +33,67 @@ const (
 type PerChainQueryResponseInternal struct {
 	RequestID  string
 	RequestIdx int
-	ChainID    vaa.ChainID
+	ChainId    vaa.ChainID
 	Status     QueryStatus
-	Results    []EthCallQueryResponse
+	Response   ChainSpecificResponse
 }
 
 // CreatePerChainQueryResponseInternal creates a PerChainQueryResponseInternal and returns a pointer to it.
-func CreatePerChainQueryResponseInternal(reqId string, reqIdx int, chainID vaa.ChainID, status QueryStatus, results []EthCallQueryResponse) *PerChainQueryResponseInternal {
+func CreatePerChainQueryResponseInternal(reqId string, reqIdx int, chainId vaa.ChainID, status QueryStatus, response ChainSpecificResponse) *PerChainQueryResponseInternal {
 	return &PerChainQueryResponseInternal{
 		RequestID:  reqId,
 		RequestIdx: reqIdx,
-		ChainID:    chainID,
+		ChainId:    chainId,
 		Status:     status,
-		Results:    results,
+		Response:   response,
 	}
 }
 
 var queryResponsePrefix = []byte("query_response_0000000000000000000|")
 
+// QueryResponsePublication is the response to a QueryRequest.
 type QueryResponsePublication struct {
 	Request           *gossipv1.SignedQueryRequest
-	PerChainResponses []PerChainQueryResponse
+	PerChainResponses []*PerChainQueryResponse
 }
 
+// PerChainQueryResponse represents a query response for a single chain.
 type PerChainQueryResponse struct {
-	ChainID   uint32
-	Responses []EthCallQueryResponse
+	// ChainId indicates which chain this query was destine for.
+	ChainId vaa.ChainID
+
+	// Response is the chain specific query data.
+	Response ChainSpecificResponse
+}
+
+// ChainSpecificResponse is the interface that must be implemented by a chain specific response.
+type ChainSpecificResponse interface {
+	Type() ChainSpecificQueryType
+	Marshal() ([]byte, error)
+	Unmarshal(data []byte) error
+	UnmarshalFromReader(reader *bytes.Reader) error
+	Validate() error
 }
 
+// EthCallQueryResponse implements ChainSpecificResponse for an EVM eth_call query response.
 type EthCallQueryResponse struct {
 	Number *big.Int
 	Hash   common.Hash
 	Time   time.Time
-	Result []byte
-	// NOTE: If you modify this struct, please update the Equal() method for QueryResponsePublication.
-}
-
-const (
-	QUERY_REQUEST_TYPE_ETH_CALL = uint8(1)
-)
 
-func (resp *QueryResponsePublication) RequestID() string {
-	if resp == nil || resp.Request == nil {
-		return "nil"
-	}
-	return hex.EncodeToString(resp.Request.Signature)
+	// Results is the array of responses matching CallData in EthCallQueryRequest
+	Results [][]byte
```

Comment:
Out of curiosity, why is Number on line 80 a big.Int when the (un)marshal treats it like a uint64?

Reply:
Changed it to use uint64 in the internal struct.